external/storm-cassandra/src/main/java/org/apache/storm/cassandra/bolt/BatchCassandraWriterBolt.java

```diff
@@ -0,0 +1,192 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+public class BatchCassandraWriterBolt extends BaseCassandraBolt<List<Tuple>> {
+
+    private final static Logger LOG = LoggerFactory.getLogger(BatchCassandraWriterBolt.class);
+
+    public static final int DEFAULT_EMIT_FREQUENCY = 2;
+
+    private static final int QUEUE_MAX_SIZE = 1000;
+
+    private LinkedBlockingQueue<Tuple> queue;
+    
+    private int tickFrequencyInSeconds;
+    
+    private long lastModifiedTimesMillis;
+
+    private String componentID;
+
+    private AsyncResultHandler<List<Tuple>> asyncResultHandler;
+
+    /**
+     * Creates a new {@link CassandraWriterBolt} instance.
+     *
+     * @param tupleMapper
+     */
+    public BatchCassandraWriterBolt(CQLStatementTupleMapper tupleMapper) {
+        this(tupleMapper, DEFAULT_EMIT_FREQUENCY);
+    }
+
+    /**
+     * Creates a new {@link CassandraWriterBolt} instance.
+     *
+     * @param tupleMapper
+     */
+    public BatchCassandraWriterBolt(CQLStatementTupleMapper tupleMapper, int tickFrequencyInSeconds) {
+        super(tupleMapper);
+        this.tickFrequencyInSeconds = tickFrequencyInSeconds;
+    }
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void prepare(Map stormConfig, TopologyContext topologyContext, OutputCollector outputCollector) {
+        super.prepare(stormConfig, topologyContext, outputCollector);
+        this.componentID = topologyContext.getThisComponentId();
+        this.queue = new LinkedBlockingQueue<>(QUEUE_MAX_SIZE);
+        this.lastModifiedTimesMillis = now();
+    }
+
+    @Override
+    protected AsyncResultHandler<List<Tuple>> getAsyncHandler() {
+        if( asyncResultHandler == null) {
+            asyncResultHandler = new BatchAsyncResultHandler(getResultHandler());
+        }
+        return asyncResultHandler;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void process(Tuple input) {
+        if( ! queue.offer(input) ) {
+            LOG.info(logPrefix() + "The message queue is full, preparing batch statement...");
+            prepareAndExecuteStatement();
+            queue.add(input);
+        }
+    }
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void tick() {
+        prepareAndExecuteStatement();
+    }
+
+    public void prepareAndExecuteStatement() {
+        int size = queue.size();
+        if( size > 0 ) {
+            List<Tuple> inputs = new ArrayList<>(size);
+            queue.drainTo(inputs);
+            try {
+                List<PairStatementTuple> psl = buildStatement(inputs);
+
+                int sinceLastModified = updateAndGetSecondsSinceLastModified();
+                LOG.debug(logPrefix() + String.format("Execute cql batches with %s statements after %s seconds", size, sinceLastModified));
```

Comment:
Using SLF4J's {} placeholders would be more elegant and also efficient; see [http://www.slf4j.org/faq.html#logging_performance]